### PR TITLE
[SYCL][RTC] Add support for errors and build log in JIT-based RTC

### PR DESCRIPTION
**Pull request by sommerlukas**
Adds support for correct error message routing and the `save_log` property to the SYCL-RTC implementation based on the SYCL-JIT library. 

The diagnostic messages associated with any errors encountered during runtime compilation is stored in the `what` of the exception thrown. 

If the user passes the `save_log` property defined in the `kernel_compiler` extension, all diagnostics encountered during compilation (errors, warnings, notes and remarks) are stored in the user-provided string.

### CODE REVIEW

**Comment by sommerlukas** [comment]
This PR currently depends on https://github.com/intel/llvm/pull/16109 for the definition of `RTCResult` and is therefore marked draft.

**Comment by sommerlukas** [comment]
@cperkinsintel @jopperm @gmlueck I'd like to hear your opinion on one aspect of the implementation. 

In the current implementation, I decided to only include the essence of the error diagnostic message in the `what` message of the exception. 

For the error in the e2e tests, this means that the `what` message is:
```
Device compilation failed
Detailed information:
error: use of undeclared identifier 'no'
error: expected ';' at end of declaration
For more information, use the 'save_log' property to obtain the full build log
```

The build log (via `save_log` property) contains the full details on the error (and any other diagnostics):
```
/home/lukas/Code/RTC/upstream/dpcpp/build/rtc.cpp:9:44: error: use of undeclared identifier 'no'
    9 |   sycl::id<1> GId = Item.get_global_id() + no semi colon !!
      |                                            ^
/home/lukas/Code/RTC/upstream/dpcpp/build/rtc.cpp:9:46: error: expected ';' at end of declaration
    9 |   sycl::id<1> GId = Item.get_global_id() + no semi colon !!
      |                                              ^
      |                                              ;
```

My thinking was that we don't want to include a verbose error message, including other non-error diagnostics, in `what` of the exception. 

If you prefer to have the full build log in `what` of the exception, I can also implement that. 

What's your preference?

File: sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp [hunk 1]
```diff
@@ -187,6 +277,10 @@ jit_compiler::compileDeviceCode(InMemoryFile SourceFile,
   FixedCompilationDatabase DB{".", CommandLine};
   ClangTool Tool{DB, {SourceFile.Path}};
 
+  IntrusiveRefCntPtr<DiagnosticOptions> DiagOpts{new DiagnosticOptions};
+  ClangDiagnosticWrapper Wrapper(BuildLog, DiagOpts.get());
+  Tool.setDiagnosticConsumer(Wrapper.consumer());
```
[the remark this answers is not in the document]
Comment (jopperm) [reply sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp:282]: Something still prints summaries, e.g. `1 warning generated.` to console, do you know how to suppress these?

File: sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp [hunk 2]
```diff
@@ -317,22 +411,20 @@ Error jit_compiler::linkDeviceLibraries(llvm::Module &Module,
   //       mandated by the extension.
```
Comment (jopperm) [inline sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp:411]: I'd argue we can drop the comment. With the new wrapper classes, this ties well into the overall RTC error reporting and I think getting the same diagnostic messages as the normal compilation flow makes it worth it.

**Comment by sommerlukas** [comment]
>  but would it be possible to include the source location in the short error messages that go into the exception, to make them more useful on their own

The `TextDiagnosticBuffer` has `SourceLocation`, but to turn them into a `FullSourceLoc`, we would need a `SourceManager`. To me, it seemed to be a bit of an overkill to set all that up.

**Comment by sommerlukas** [comment]
> > In the current implementation, I decided to only include the essence of the error diagnostic message in the what message of the exception.
> > My thinking was that we don't want to include a verbose error message, including other non-error diagnostics, in `what` of the exception.
> 
> The [specification](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc#new-free-functions-to-create-and-build-kernel-bundles) addresses this:
> 
> > An exception with the `errc::build` error code if the compilation or linking operations fail. In this case, the exception `what` string provides a full build log, including descriptions of any errors, warning messages, and other diagnostics. This string is intended for human consumption, and the format may not be stable across implementations of this extension.
> 
> I found that this was pretty convenient when developing test applications. I didn't need to set up an exception handler or use the `build_log` property. The default behavior in this case is to terminate the application and print a nice error message that tells me exactly what problem I had in my source string.

That's a good point, it's probably more convenient to get the full error message right away than having to go back, add the `save_log` property and recompile and rerun. 

> Of course, we can change the specification if we think the other behavior is better, but I wonder how easy it is to determine the "essence" of the error. Quite often, a mistake in C++ syntax will lead to a raft of error messages, and I think it is not an easy task to figure out which of these is the essence of the problem.
> 
> I also suspect that changing the specification in this way would make the extension more difficult for other vendors to implement. The way it is written now, a vendor can spawn the compiler in a separate process and capture the stdout / stderr into the `what` string. If a vendor wants to implement the extension by spawning the compiler in a separate process, would it have to parse through the stderr output to find the essence of the error?

So far, I'm simply using the severity of the diagnostic given by Clang/LLVM to decide what should be in `what`. But you're right, if the compiler does not separate errors and warnings into different streams (e.g., `stderr` and `stdout`) or has flags to achieve such behavior, it would be hard for other implementations if we changed the specification. 

Based on your points and @jopperm's feedback that source locations would be nice, I'll change the implementation to simply include the entire build log in `what` (also simplifies our implementation).

I'll update the PR accordingly, thanks for the feedback!

**Comment by jopperm** [comment]
I addressed the CI errors in #16109, so these should be gone after rebasing.

File: sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (sommerlukas) [reply sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp:257]: Summaries should not be printed anymore after the latest changes.

File: sycl-jit/jit-compiler/include/KernelFusion.h [hunk 3]
```diff
@@ -61,13 +61,13 @@ class RTCResult {
   explicit RTCResult(const char *ErrorMessage)
       : Failed{true}, BundleInfo{}, ErrorMessage{ErrorMessage} {}
 
-  explicit RTCResult(RTCBundleInfo &&BundleInfo)
-      : Failed{false}, BundleInfo{std::move(BundleInfo)}, ErrorMessage{} {}
+  explicit RTCResult(RTCBundleInfo &&BundleInfo, const char *BuildLog)
+      : Failed{false}, BundleInfo{std::move(BundleInfo)}, ErrorMessage{
+                                                              BuildLog} {}
```
Comment (victor-eds) [inline sycl-jit/jit-compiler/include/KernelFusion.h]:
```suggestion
  RTCResult(RTCBundleInfo &&BundleInfo, const char *BuildLog)
      : Failed{false}, BundleInfo{std::move(BundleInfo)}, ErrorMessage{
                                                              BuildLog} {}
```

Drop `explicit`

File: sycl-jit/jit-compiler/include/KernelFusion.h [hunk 4]
```diff
@@ -61,13 +61,13 @@ class RTCResult {
...
 
   bool failed() const { return Failed; }
 
   const char *getErrorMessage() const {
-    assert(failed() && "No error message present");
```
Comment (victor-eds) [inline sycl-jit/jit-compiler/include/KernelFusion.h:70]: Why are we dropping this?

File: sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp [hunk 5]
```diff
@@ -161,12 +172,56 @@ struct GetLLVMModuleAction : public ToolAction {
...
+  llvm::raw_string_ostream LogStream;
+
+  std::unique_ptr<clang::TextDiagnosticPrinter> LogPrinter;
+
+public:
+  ClangDiagnosticWrapper(std::string &LogString, DiagnosticOptions *DiagOpts)
+      : LogStream(LogString) {
+
+    LogPrinter = std::make_unique<TextDiagnosticPrinter>(LogStream, DiagOpts);
+  }
```
Comment (victor-eds) [inline sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp:186]:
```suggestion
  ClangDiagnosticWrapper(std::string &LogString, DiagnosticOptions *DiagOpts)
      : LogStream(LogString), LogPrinter(std::make_unique<TextDiagnosticPrinter>(LogStream, DiagOpts)) {}
```

Comment (victor-eds) [reply sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp:186]: Initialization order is ensured.

File: sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.h [hunk 6]
```diff
@@ -17,15 +17,18 @@
...
 #include <memory>
+#include <string>
 
 namespace jit_compiler {
 
 llvm::Expected<std::unique_ptr<llvm::Module>>
 compileDeviceCode(InMemoryFile SourceFile, View<InMemoryFile> IncludeFiles,
-                  const llvm::opt::InputArgList &UserArgList);
+                  const llvm::opt::InputArgList &UserArgList,
+                  std::string &BuildLog);
```
Comment (victor-eds) [inline sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.h:27]: Can we use `std::string` in this header for ABI purposes?

File: sycl-jit/jit-compiler/include/KernelFusion.h (hunk 4 above)
Comment (sommerlukas) [reply sycl-jit/jit-compiler/include/KernelFusion.h:70]: We use the error message field for the build log when compilation was successful, so we need to be able to access this field even if the result isn't failed.

File: sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.h (hunk 6 above)
Comment (sommerlukas) [reply sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.h:27]: This header is internal to the JIT library, so no ABI concerns here.

File: sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp [hunk 7]
```diff
@@ -197,12 +252,19 @@ jit_compiler::compileDeviceCode(InMemoryFile SourceFile,
...
+
   // Set up in-memory filesystem.
   Tool.mapVirtualFile(SourceFile.Path, SourceFile.Contents);
   for (const auto &IF : IncludeFiles) {
     Tool.mapVirtualFile(IF.Path, IF.Contents);
   }
 
+  // Suppress message "Error while processing" being printed to stdout.
+  Tool.setPrintErrorMessage(false);
+
```
Comment (jopperm) [inline sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp:267]: Super nit: This feels related to the diagnostics setup, so move before VFS setup.

File: sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp [hunk 8]
```diff
@@ -161,12 +172,55 @@ struct GetLLVMModuleAction : public ToolAction {
...
+  bool handleDiagnostics(const DiagnosticInfo &DI) override {
+    auto Prefix = [](DiagnosticSeverity Severity) -> llvm::StringLiteral {
+      switch (Severity) {
+      case llvm::DiagnosticSeverity::DS_Error:
+        return "ERROR";
+      case llvm::DiagnosticSeverity::DS_Warning:
+        return "WARNING";
+      default:
+        return "NOTE:";
+      }
```
Comment (victor-eds) [inline sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp:210]:
```suggestion
      switch (Severity) {
      case llvm::DiagnosticSeverity::DS_Error:
        return "ERROR";
      case llvm::DiagnosticSeverity::DS_Warning:
        return "WARNING";
      case llvm::DiagnosticSeverity::DS_Note:
        return "NOTE:";
      case llvm::DiagnosticSeverity::DS_Remark:
        return "REMARK:";
      default:
        llvm_unreachable("Unhandled case");
      }
```

Better?

**Review by victor-eds — COMMENTED** [review]
LGTM but NIT.

**Review by jopperm — APPROVED** [review]

**Review by victor-eds — APPROVED** [review]

**Review by cperkinsintel — APPROVED** [review]

**Comment by sommerlukas** [comment]
Post-commit failure is addressed in https://github.com/intel/llvm/pull/16243.

**Pull request merged**